### ISSUE TITLE
Submission Username Customization

### ISSUE THREAD

**Issue by whitead** [jupyter/nbgrader#667]
I use jupyterhub with a dockerspawner so that each jupyter notebook server is run in a separate container. All containers have a single user called `jovyan` so that when `get_username()` is called for submitting an assignment, it's always `jovyan`. It would be nice if there was a way to change that. Jupyterhub deals with this with the user traitlet and dockerspawner has an environment variable (`JPY_USER`) set. 

Any ideas?

**Comment by whitead** [comment]
One idea might be to check for an environment variable first, like `NBG_USER`:

```python
def get_username():
    """Get the username from environment variable NBG_USER or of the current process if NBG_USER is not set."""
    if pwd is None:
        raise OSError("get_username cannot be called on Windows")

    jpy_uname = os.environ.get('NBG_USER')
    if jpy_uname:
        return jpy_uname
    return pwd.getpwuid(os.getuid())[0]
```

This would require rewriting the `self_owned` method to:

```python
def self_owned(path):
    """Is the path owned by the current user of this process?"""
    return pwd.getpwuid(os.getuid())[0] == find_owner(os.path.abspath(path))
```

**Comment by jhamrick** [comment]
Yeah, the current way that the exchange directory is set up doesn't deal with this use case particularly well. The problem with adding this option is that it would then allow *any* student to submit as any other student, because the way the permissions work is based on user id. And, if you were releasing assignments from your JupyterHub instance (i.e., if your instructor account is also the jovyan user) then students would also be able to overwrite your release and collect the submissions from other students. Regardless, with or without this option, students would still be able to manually go to the exchange directory and read or modify any files there if they all have the same user id. So, I really don't think it's a good idea to set it up with everybody having the same user id in the first place.

In my case, I used `SystemUserSpawner` (which is basically just `DockerSpawner`, except it supports creating different users in the different docker containers) and created a Dockerfile that inherited from the  `systemuser` Dockerfile (https://github.com/jupyterhub/dockerspawner/tree/master/systemuser). I would recommend taking this approach rather than adding an option to set the username based on an environment variable.

We have in the works a better way to do this (see #659) but it's unfortunately not yet implemented.

**Comment by whitead** [comment]
Thank you for the detailed reply. I understand the security issue of students changing the environment variables, but in your example what prevents students from just modifying the /srv/exchange directory since it has 777 permissions? A student could just using the bash magic or python shutil to accomplish the same task, right? I assumed integrity of the submission process was a separate issue than the students submission....

**Comment by jhamrick** [comment]
The exchange directory has carefully crafted permissions to prevent students from accessing and modifying each others' assignments. It's not well document, but hopefully we won't need such a complicated solution soon anyway 😄 

Here's the detailed explanation of how it works:

The toplevel exchange directory has 777 permissions, but the subdirectories have different permissions. The inbound directory is writable by all, but not readable or listable by anybody but the instructor. This has the effect that students can write to the directory in order to submit their assignments, but cannot see anybody else's submissions. The outbound directory is readable by all but not writable by anybody but the instructor, so that only the instructor can release assignments:

```
$ ls -l /srv/nbgrader/exchange/cogsci131/
total 272
drwx-ws-wx 3968 instructor instructor 270336 Jan 17 04:22 inbound
drwxr-xr-x   21 instructor instructor   4096 Dec  5 16:56 outbound
```

Submission *directories* in the inbound directory are readable and writeable by everybody (though note because the inbound directory is not readable by anyone except the instructor, no one but the instructor can actually read or list these directories):

```
$ ls -l /srv/nbgrader/exchange/cogsci131/inbound/
total 15864
drwxrwxrwx 4 student1 instructor 4096 Dec 12 07:46 student1+final+2016-12-12 07:46:33 UTC
```

But the individual files within the submission directories are writable only by the student, which means that even if another student tried to remove the submission directory, they would not have permission to remove the files inside it:

```
$ ls -l /srv/nbgrader/exchange/cogsci131/inbound/student1+final+2016-12-12\ 07\:46\:33\ UTC/
total 524
-rw-r--r-- 1 student1 instructor   3946 Dec  5 16:56 0 - Index.ipynb
...
```

Released assignments are only writable by the instructor, but readable by everybody:

```
$ ls -l /srv/nbgrader/exchange/cogsci131/outbound/
total 76
drwxr-xr-x 4 instructor instructor 4096 Dec  5 10:23 final
$ ls -l /srv/nbgrader/exchange/cogsci131/outbound/final/
total 144
-rw-r--r-- 1 instructor instructor  3946 Dec  5 16:56 0 - Index.ipynb
```

**Comment by whitead** [comment]
Wow, alright. It sounds like you've really thought this through. I hope there is a way though to customize the transferApp later to accommodate other installations. Thank you.

**Comment by jhamrick** [comment]
Yes, as I mentioned we are definitely planning on redoing this part of nbgrader so it doesn't depend on a local, shared directory (see #659 ). Sorry that it doesn't work the installation you had in mind at the moment!

**Comment by nthiery** [comment]
For reference: student id customization was implemented in #1098.
